boa_engine/src/builtins/promise/promise_job.rs

```diff
@@ -0,0 +1,194 @@
+    pub(crate) fn new_promise_resolve_thenable_job(
+        promise_to_resolve: JsObject,
+        thenable: JsValue,
+        then: JobCallback,
+        context: &mut Context,
+    ) -> JobCallback {
+        // 1. Let job be a new Job Abstract Closure with no parameters that captures promiseToResolve, thenable, and then and performs the following steps when called:
+        let job = FunctionBuilder::closure_with_captures(
+            context,
+            |this: &JsValue, args: &[JsValue], captures, context: &mut Context| {
+                let JobCapture {
+                    promise_to_resolve,
+                    thenable,
+                    then,
+                } = captures;
+
+                //    a. Let resolvingFunctions be CreateResolvingFunctions(promiseToResolve).
+                let resolving_functions =
+                    Promise::create_resolving_functions(promise_to_resolve, context)?;
+
+                //    b. Let thenCallResult be Completion(HostCallJobCallback(then, thenable, « resolvingFunctions.[[Resolve]], resolvingFunctions.[[Reject]] »)).
+                let then_call_result = then.call_job_callback(
+                    thenable,
+                    &[
+                        resolving_functions.resolve,
+                        resolving_functions.reject.clone(),
+                    ],
+                    context,
+                );
+
+                //    c. If thenCallResult is an abrupt completion, then
+                if let Err(value) = then_call_result {
+                    //    i. Return ? Call(resolvingFunctions.[[Reject]], undefined, « thenCallResult.[[Value]] »).
+                    return context.call(
+                        &resolving_functions.reject,
+                        &JsValue::Undefined,
+                        &[value],
+                    );
+                }
+
+                //    d. Return ? thenCallResult.
+                then_call_result
+            },
+            JobCapture::new(promise_to_resolve, thenable, then),
+        )
+        .build();
+
+        // 2. Let getThenRealmResult be Completion(GetFunctionRealm(then.[[Callback]])).
+        // 3. If getThenRealmResult is a normal completion, let thenRealm be getThenRealmResult.[[Value]].
+        // 4. Else, let thenRealm be the current Realm Record.
+        // 5. NOTE: thenRealm is never null. When then.[[Callback]] is a revoked Proxy and no code runs, thenRealm is used to create error objects.
+        // 6. Return the Record { [[Job]]: job, [[Realm]]: thenRealm }.
+        JobCallback::make_job_callback(job.into())
+    }
+}
+
+#[derive(Debug, Trace, Finalize)]
+struct JobCapture {
+    promise_to_resolve: JsObject,
+    thenable: JsValue,
+    then: JobCallback,
+}
+
+impl JobCapture {
+    fn new(promise_to_resolve: JsObject, thenable: JsValue, then: JobCallback) -> Self {
+        Self {
+            promise_to_resolve,
+            thenable,
+            then,
+        }
+    }
+}
```

Comment:
This would need some documentation to know where it's coming from, how to use it, what it means and so on. Maybe even a spec link. Also for the `new()` function.